packages/features/ee/billing/credit-service.ts

```diff
@@ -30,100 +65,132 @@ export class CreditService {
+          await this._handleLowCreditBalanceResult(result.lowCreditBalanceResult);
+        }
+        return {
+          teamId: result?.teamId,
+          userId: result?.userId,
+        };
+      });
   }
 
   async hasAvailableCredits({ userId, teamId }: { userId?: number | null; teamId?: number | null }) {
-    if (!IS_SMS_CREDITS_ENABLED) return true;
-
-    if (teamId) {
-      const creditBalance = await CreditsRepository.findCreditBalance({ teamId });
-
-      const limitReached =
-        creditBalance?.limitReachedAt &&
-        dayjs(creditBalance.limitReachedAt).isAfter(dayjs().startOf("month"));
-
-      if (!limitReached) return true;
-
-      // check if team is still out of credits
-      const teamCredits = await this.getAllCreditsForTeam(teamId);
-      const availableCredits = teamCredits.totalRemainingMonthlyCredits + teamCredits.additionalCredits;
-
-      if (availableCredits > 0) {
-        await CreditsRepository.updateCreditBalance({
-          teamId,
-          data: {
-            limitReachedAt: null,
-            warningSentAt: null,
-          },
-        });
-        return true;
+    return await prisma.$transaction(async (tx) => {
+      if (!IS_SMS_CREDITS_ENABLED) return true;
+
+      if (teamId) {
+        const creditBalance = await CreditsRepository.findCreditBalance({ teamId }, tx);
+
+        const limitReached =
+          creditBalance?.limitReachedAt &&
+          dayjs(creditBalance.limitReachedAt).isAfter(dayjs().startOf("month"));
+
+        if (!limitReached) return true;
+
+        // check if team is still out of credits
+        const teamCredits = await this._getAllCreditsForTeam({ teamId, tx });
+        const availableCredits = teamCredits.totalRemainingMonthlyCredits + teamCredits.additionalCredits;
+
+        if (availableCredits > 0) {
+          await CreditsRepository.updateCreditBalance(
+            {
+              teamId,
+              data: {
+                limitReachedAt: null,
+                warningSentAt: null,
+              },
+            },
+            tx
+          );
+          return true;
+        }
       }
-    }
 
-    if (userId) {
-      const teamWithAvailableCredits = await this.getTeamWithAvailableCredits(userId);
+      if (userId) {
+        const teamWithAvailableCredits = await this._getTeamWithAvailableCredits({ userId, tx });
 
-      if (teamWithAvailableCredits && teamWithAvailableCredits?.availableCredits > 0) return true;
+        if (teamWithAvailableCredits && teamWithAvailableCredits?.availableCredits > 0) return true;
 
-      const userCredits = await this.getAllCredits({ userId });
+        const userCredits = await this._getAllCredits({ userId, tx });
 
-      return userCredits.additionalCredits > 0;
-    }
+        return userCredits.additionalCredits > 0;
+      }
 
-    return false;
+      return false;
+    });
   }
 
   async getTeamWithAvailableCredits(userId: number) {
```

Comment:
`getTeamWithAvailableCredits` is used externally and internally

- externally: we call `await creditService.getTeamWithAvailableCredits(...)`
- internally: Within `credit-service.ts`, we call this method as a part of a transaction, for example, in `hasAvailableCredits`.

So I've made two versions:

- `getTeamWithAvailableCredits`: public method, running $transaction and pass `tx` to the protected version
- `_getTeamWithAvailableCredits`: protected method, getting `tx` as a parameter